discover/server.go

```
@@ -0,0 +1,97 @@
+package discover
+
+import (
+	"github.com/coreos/go-etcd/etcd"
+	"github.com/flynn/rpcplus"
+	"log"
+	"net"
+	"net/http"
+)
+
+const (
+	HeartbeatIntervalSecs = 5
+	MissedHearbeatTTL     = 5
+)
+
+type ServiceUpdate struct {
+	Name   string
+	Addr   string
+	Online bool
+	Attrs  map[string]string
+}
+
+type Args struct {
+	Name  string
+	Addr  string
+	Attrs map[string]string
+}
+
+type DiscoveryBackend interface {
+	Subscribe(string) (chan *ServiceUpdate, error)
+	Register(string, string, map[string]string) error
+	Unregister(string, string) error
+	Heartbeat(string, string) error
+}
+
+type DiscoverAgent struct {
+	Backend DiscoveryBackend
+	Address string
+}
+
+func NewServer() *DiscoverAgent {
+	return &DiscoverAgent{
+		Backend: &EtcdBackend{Client: etcd.NewClient()},
+		Address: ":1111",
+	}
+}
+
+func ServeForever(server *DiscoverAgent) {
```

The idiomatic name for this function is `ListenAndServe` and it should return an error if `Listen` fails.